### ISSUE TITLE
hanging parsing, slowing down

### ISSUE THREAD

**Issue by chrmarti** [microsoft/vscode-cpptools#1118]
_From @lukaswajdzik on October 6, 2017 11:42_

- VSCode Version: Code 1.17.0 (be377c0faf7574a59f84940f593a6849f12e4de7, 2017-10-04T23:33:45.303Z)
- OS Version: Linux x64 4.10.0-35-generic
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
include-autocomplete|ajs|0.0.4
code-gnu-global|aus|0.2.2
gitlens|eam|5.4.1
CppSnippets|har|0.0.12
vscode-clang|mit|0.2.2
cpptools|ms-|0.13.1
vscode-icons|rob|7.14.0

---

Steps to Reproduce:

1. Open c++ project (>10 000 files)
2. Working very slow, indexing cannot finish

Previously I had no such a problem (previous versions). It was parsing quickly and working fast

_Copied from original issue: Microsoft/vscode#35712_

**Comment by chrmarti** [comment]
Could you check if there is a `rg` process running and what arguments that has?

Could you also post your `files.exclude` and `search.exclude` settings here?

**Comment by chrmarti** [comment]
_From @lukaswajdzik on October 9, 2017 7:48_

thanks for quick reply!
I don't see any rg process running (_ps -aux | grep rg_)
    "files.exclude": {
        "**/.git": true,
        "**/.svn": true,
        "**/.hg": true,
        "**/CVS": true,
        "**/.DS_Store": true,
        "**/cmake-build-debug": true
    }
  "search.exclude": {
    "**/node_modules": true,
    "**/bower_components": true
  }

**Comment by chrmarti** [comment]
_From @Villentre on October 9, 2017 9:2_

Same issue here. 
In C++ project parsing run all the time (parsing-fire icon) and CPU is heavily used.

I have revert `ms-vscode.cpptools` extension to version `0.12.4` and it seems to be ok.

**Comment by chrmarti** [comment]
_From @lukaswajdzik on October 9, 2017 9:15_

It helps! Thank you.

So the issue is with c++ extension.

**Comment by bobbrow** [comment]
@lukaswajdzik and @Villentre, can you give us more information about your projects?  Are they open source so we can debug locally?  Can you share your c_cpp_properties.json file?

Can you be more specific about what you mean by "working very slow"?  Do you just mean that the indexing is taking longer than you expect?

**Comment by sean-mcmanus** [comment]
@bobbrow I have a suspicion as to what is happening. With 0.13.1, we fixed a bug where the headers were not being re-parsed if an abort occurred while parsing, so if users had entered an infinite loop during parsing of a file, they would skip parsing that file the next time, so the issue would seem fixed (when in reality some files were never being parsed). We really need logs and/or a repro to fix this, but as a workaround we could "incorrectly" mark the file as parsed if it takes longer than a minute or so (it's possible we could abort too, but without a repro it's hard to tell if that would work).

**Comment by Villentre** [comment]
@bobbrow Sorry, I have missed your post.
@sean-mcmanus Thanks for bringing it up!

In my case it's non-open source repo, so I can't share it with you. 
It is split it into components, so I open each component separately in vscode (about 5k hpp/cpp files).

I'm using `"C_Cpp.intelliSenseEngine": "Tag Parser"` and here's my config:
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "/usr/include/x86_64-linux-gnu/c++/6",
                "/usr/include/c++/6",
                "/usr/local/include",
                "/usr/include",
                "${workspaceRoot}"
            ],
            "defines": [],
            "intelliSenseMode": "clang-x64",
            "browse": {
                "path": [
                    "/usr/include/x86_64-linux-gnu/c++/6",
                    "/usr/include/c++/6",
                    "/usr/local/include",
                    "/usr/include",
                    "/work/generated", // generated files with i.e. struct definitions
                    "/work/externals", // googletest, boost, ...
                    "${workspaceRoot}" // /work/project
                ],
                "limitSymbolsToIncludedHeaders": true,
                "databaseFilename": ""
            }
        }
    ],
    "version": 3
}
```

I'll try to collect logs form version 0.12.4 (when I'm sure, all works fine) and version 0.13.1.

**Comment by sean-mcmanus** [comment]
@Villentre Using `/usr/include/x86_64-linux-gnu/c++/6` and `/usr/include/c++/6` in your browse.path will cause unnecessary processing because it's already covered by `/usr/include` (it's required for the includePath), but that will be fixed in our next update.

We don't necessarily need your source code, but via using logs and removing unnecessary code it may be possible to generate a generic code sample repro. Also, if you can identify the offending files or folders that cause the infinite processing, you could add them to the files.exclude as a workaround to unblock you. A first step would be to remove all the folders from browse.path, and verify that it stops processing and then add folders one by one (or in halves) until the folders are found that cause the bug.

**Comment by lukaswajdzik** [comment]
Hello, please see my log file from version 0.13.1.
Is seems that the indexer simply failed.
I substituted some direct file paths with *path*

[vscode.cpp.log.19797.txt](https://github.com/Microsoft/vscode-cpptools/files/1381634/vscode.cpp.log.19797.txt)

**Comment by sean-mcmanus** [comment]
@lukaswajdzik Yeah, tag_parse failed is occurring. That will cause the particular file to not have symbols parsed. Are you seeing other buggy symptoms? If you could help identify a generic type of code that is causing this we could fix it.

**Comment by sean-mcmanus** [comment]
@lukaswajdzik @Villentre We have a pre-release with the fix if you want to try it out (https://github.com/Microsoft/vscode-cpptools/releases/tag/v0.14.6-insiders) and let us know if there are any fixes we should make before our release next week.

**Comment by lukaswajdzik** [comment]
It looks good! Seem that previous issue is fixed. Thanks

2018-01-10 0:33 GMT+01:00 Sean McManus <notifications@github.com>:

**Comment by sean-mcmanus** [comment]
@lukaswajdzik Yeah, sorry for taking so long to fix that -- we didn't realize it was a general wide spread regression due to all the normal special case issues we hear about. It wasn't until a user reported the \u repro that we were able to repro and debug it.

**Comment by sean-mcmanus** [comment]
Fixed. https://github.com/Microsoft/vscode-cpptools/releases/latest
